src/Scheduler.scala

```diff
@@ -0,0 +1,32 @@
+import java.util.concurrent.{Executors, PriorityBlockingQueue}
+
+import scala.concurrent.{ExecutionContext, Future}
+
+trait Scheduler {
+  def schedule[T](task: () => T, timestamp: Long): Future[T]
+  def start()(implicit ec: ExecutionContext): Unit
+}
+
+
+class SchedulerImp(val minHeap: PriorityBlockingQueue[ScheduledTask[_]]) extends Scheduler {
+
+  // ExecutionContext for Future
+  implicit val ec: ExecutionContext = ExecutionContext.fromExecutorService(Executors.newFixedThreadPool(5))
+
+  override def schedule[T](task: () => T, timestamp: Long): Future[T] = {
+    minHeap.offer(ScheduledTask[T](task, timestamp))
+    Future {
```

Comment:
The task should be executed when it's taken out from the queue and not in here. To return `Future[T]` we need to use `Promise` and add the promise to the queue to keep the reference, so that we can get the result back when it's executed in the future.